src/hash.c

```diff
@@ -29,11 +29,11 @@ static uint32_t hash_function(const void *p, size_t len) {
 	uint32_t hash = 0;
 
 	while(true) {
-		for(int i = len > 4 ? 4 : len; --i;) {
+		for(size_t i = len > 4 ? 4 : len; --i;) {
 			hash += (uint32_t)q[len - i] << (8 * i);
 		}
 
-		hash *= 0x9e370001UL; // Golden ratio prime.
+		hash = (uint32_t)(hash * 0x9e370001UL); // Golden ratio prime.
```

Comment:
Hm, this should probably just be using the `U` instead of `UL` postfix:

```
hash *= 0x9e370001U;
```

Reply:
This I wasn't able to change. If we [replace](https://github.com/hg/tinc/runs/3155144177) this line with

```c
hash *= 0x9e370001U; // Golden ratio prime.
```

or even with an [explicit cast](https://github.com/hg/tinc/runs/3155098787) after the fact

```c
hash = (uint32_t)(hash * 0x9e370001U); // Golden ratio prime.
```

it still does not make the sanitizer happy:

    hash.c:36:8: runtime error: unsigned integer overflow: 512 * 2654404609 cannot be represented in type 'unsigned int'
        #0 0x55ea5adceffe in hash_function src/hash.c:36:8
        #1 0x55ea5adcf166 in hash_search src/hash.c:86:22
        #2 0x55ea5ae6e8b6 in lookup_subnet_ipv6 src/subnet.c:192:10
        ...

It seems to be very strict with overflows and forces us to make them explicit. We can always [weaken the checks](https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html#silencing-unsigned-integer-overflow), of course.

Reply:
Since it's just this one one thing, let's keep this check enabled.